docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,233 @@
+
+### Source Generator Integration
+
+An example of how the previous P/Invoke snippet could be transformed is below. This example is using the proposed API in this document. The Source Generator currently has a limitation that we will ignore at present - the inability to modify user written code. This limitation is ignored to present an optimal user experience, but will be discussed later in case the current limitation becomes a hard requirement.
+
+`Program.cs` (Pre-generated code)
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* D */ [GenerateNativeImportAttribute]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C*/ QueryPerformanceCounter(out count);
+```
+
+Observe point (D), the new attribute. This attribute provides an indication to a Source Generator that the following declaration should be updated.
+
+`Program.cs` (In-memory updated code)
+
+``` CSharp
+/* D */ [GenerateNativeImportAttribute]
+/* B */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+During the source generation process the `DllImportAttribute` (A) would be removed. The `GenerateNativeImportAttribute` (D) would remain to provide a metadata indication that the function was auto-generated. Also note that the method declaration has been updated and marked `partial`. The Source Generator would then generate the source for this partial method. The invocation (C) remains unchanged.
+
+`Stubs.g.cs`:
+
+``` CSharp
+/* E */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount)
+{
+    unsafe
+    {
+        long result = 0;
+        bool success = QueryPerformanceCounter(&result) != 0;
+        lpPerformanceCount = result;
+        return success;
+    }
+}
+
+[DllImportAttribute("Kernel32.dll")]
+/* F */ private static extern int QueryPerformanceCounter(long* lpPerformanceCount);
+```
+
+The Source Generator would generate the implementation of the partial method (E) in a separate translation unit (`Stubs.g.cs`). At point (F) the `DllImportAttribute` declaration from the user's original declaration is copied into a private P/Invoke specifically for the generated code. The P/Invoke signature from the original declaration would be modified to contain only [blittable types][blittable_link] to ensure the JIT could inline the invocation. Finally note that the user's original function signature would remain in to avoid impacting existing callsites.
+
+In this system it is not defined how marshaling of specific types would be performed. The built-in runtime has complex rules for some types, and it is these rules that once shipped become the de facto standard - often times regardless if the behavior is a bug or not. The design here is not concerned with how the arguments go from a managed to unmanaged environment.
```

Comment:
It may be useful to give examples of new types of marshaling rules that this will enable.